### PR TITLE
Fast TVguide loading

### PR DESCRIPTION
**Pull request by candry7731**
Loads first 10 channels in the "guide view". Once user scrolls past the 6th channel the guide will load the next 10 channels until there is nothing left to load. 

This does not change the way the LiveTV "channel view" operates 

**Changes**
Added a new function to the schedule.brs file, advanceGuide()
Set the LoadChannelsTask.xml Limit to 10

### CODE REVIEW

**Comment by candry7731** [comment]
let me know what you guys think of this

**Comment by 1hitsong** [comment]
Focus was lost immediately after I loaded the Live TV library. My default is Guide view. If I changed to Channel View, then back to Guide, it worked.
https://user-images.githubusercontent.com/3330318/210137682-5ce93513-51ec-48dc-b7b9-ae6952709302.mp4

**Comment by 1hitsong** [comment]
Let's increase the look ahead a bit so it attempts loading the new channels off screen. 

As I scroll to the bottom of the loaded data, Channel 1 will show at the bottom, then get replaced. It jumps around when it happens, so best to do that off screen.

File: components/liveTv/schedule.brs [hunk 1]
```diff
@@ -28,55 +30,73 @@ sub init()
...
+
+    'Set initial channels loaded int - TODO create user settings to control the limi
+    m.top.channelsLoaded = 0
+
 end sub
 
 sub channelFilterSet()
-    m.scheduleGrid.jumpToChannel = 0
+    m.spinner.visible = true
+    'm.scheduleGrid.jumpToChannel = 0
```
Comment (1hitsong) [inline components/liveTv/schedule.brs]: Remove

File: components/liveTv/schedule.brs [hunk 2]
```diff
@@ -28,55 +30,73 @@ sub init()
...
+        'set guide focus
+        m.scheduleGrid.setFocus(true)
     end if
 
 end sub
 
 'Voice Search set
 sub channelsearchTermSet()
-    m.scheduleGrid.jumpToChannel = 0
+    'm.scheduleGrid.jumpToChannel = 0
```
Comment (1hitsong) [inline components/liveTv/schedule.brs]: Remove

File: components/liveTv/schedule.brs [hunk 3]
```diff
@@ -87,15 +107,16 @@ sub onChannelsLoaded()
             m.top.signalBeacon("EPGLaunchComplete") ' Required Roku Performance monitoring
             m.EPGLaunchCompleteSignaled = true
         end if
-        m.LoadChannelsTask.channels = []
-
+        'm.LoadChannelsTask.channels = m.LoadChannelsTask.channels
```
Comment (1hitsong) [inline components/liveTv/schedule.brs]: Remove

File: components/liveTv/schedule.brs [hunk 4]
```diff
@@ -169,6 +191,24 @@ sub onProgramDetailsLoaded()
...
+    if m.top.channelChanged > (m.top.channelsLoaded - 6)
+        'add loaded channels count to advance guide
+        m.top.channelsLoaded = m.top.channelsLoaded + m.LoadChannelsTask.limit
+        'advance start index to get new channels
+        startIdex = m.top.channelsLoaded - 9
+        'update satrt index in loaded channels task
+        m.LoadChannelsTask.startIndex = startIdex
+        ' if task is running stop and load more
+        m.top.counter = m.top.channelsLoaded - 11
+        'm.top.currentChannelFocused = m.scheduleGrid.channelFocused
```
Comment (1hitsong) [inline components/liveTv/schedule.brs]: Remove

File: components/movies/MovieDetails.brs [hunk 5]
```diff
@@ -19,10 +19,13 @@ sub init()
     m.buttonGrp.setFocus(true)
     m.top.lastFocus = m.buttonGrp
 
+    m.spinner = m.top.findNode("spinner")
```
Comment (1hitsong) [inline components/movies/MovieDetails.brs]: Why are there changes to MovieDetails.brs and MovieDetails.xml?

File: source/Main.brs [hunk 6]
```diff
@@ -307,16 +311,19 @@ sub Main (args as dynamic) as void
...
                     sceneManager.callFunc("pushScene", video)
+                    dialog.close = true
                 end if
 
                 if group.lastFocus <> invalid
                     group.lastFocus.setFocus(true)
                 end if
             else if btn <> invalid and btn.id = "trailer-button"
+                dialog = createObject("roSGNode", "ProgressDialog")
+                dialog.title = tr("Loading trailer")
```
Comment (1hitsong) [inline source/Main.brs:415]: Are any of these changes in Main.brs for Live TV? If not, they need to be moved out.

**Review by 1hitsong — CHANGES_REQUESTED** [review]

File: source/Main.brs (hunk 6 above)
Comment (candry7731) [reply source/Main.brs:415]: I did not add any of this trailer-button code.... not sure where it came from

**Comment by candry7731** [comment]
> Focus was lost immediately after I loaded the Live TV library. My default is Guide view. If I changed to Channel View, then back to Guide, it worked. https://user-images.githubusercontent.com/3330318/210137682-5ce93513-51ec-48dc-b7b9-ae6952709302.mp4

Hmm, I do not get this same issue. do you have guide data? I wonder if the focus is lost because the program data is never loaded.

**Comment by 1hitsong** [comment]
I clicked on Live TV, and it just spins. The guide never loads. I don't have up-to-date program data. No errors in console.

image

**Comment by 1hitsong** [comment]
I'm still having issues with focus being lost. It seems to happen when I load the TV in guide view, press back, then load TV again.

https://user-images.githubusercontent.com/3330318/216189512-c247de83-896f-4348-940d-f53a84339e4b.mp4

**Comment by candry7731** [comment]
I do not get this at all. I always have focus even when switching views and pressing back buttons. I'll try to post a small video it said mine is too big....


https://user-images.githubusercontent.com/83685828/216214999-f5fe60a7-d2f5-4c33-8971-6eb75afde340.mp4

**Comment by 1hitsong** [comment]
1. Scrolling up from channel 1 makes it freak out
2. If you do a voice search for something that returns no results, you get an endless waiting spinner.

**Comment by 1hitsong** [comment]
- [ ] Scrolling up from channel 1 makes it freak out
- [x] If you do a voice search for something that returns no results, you get an endless waiting spinner.

**Comment by 1hitsong** [comment]
I'm unsure how best to approach item 1. I feel as if disabling scrolling Up all together isn't the right solution.

Just brainstorming here, what if we load the initial 20 channel data like you have it, then trigger the load task for the next 20. Once that batch loads, we load the next 20, etc. until all channels are loaded.

If a user scrolls up, they'll see the channel, and after a moment of waiting, the data will load. Users who only scroll down will have the best experience, and users who scroll up will have only a slight delay.

Thoughts?

**Comment by candry7731** [comment]
I like that theory! Can you implement it?

**Comment by 1hitsong** [comment]
> I like that theory! Can you implement it?

I'll see what I can come up with this weekend.

**Pull request closed without merging**